server/storage/mvcc/index_test.go

```diff
@@ -235,101 +235,444 @@ func TestIndexRevision(t *testing.T) {
+					},
+				},
+				{
+					key:      []byte("foo2"),
+					modified: Revision{Main: 8},
+					generations: []generation{
+						{ver: 3, created: Revision{Main: 3}, revs: []Revision{Revision{Main: 4}, Revision{Main: 8}}},
+						{},
+					},
+				},
+			},
+			keep: map[Revision]struct{}{
+				Revision{Main: 4}: {},
+				Revision{Main: 5}: {},
+			},
+			compacted: map[Revision]struct{}{
+				Revision{Main: 7}: {},
+				Revision{Main: 4}: {},
+				Revision{Main: 5}: {},
+			},
+		},
+		{
+			atRev: 8,
+			keyIndexes: []keyIndex{
+				{
+					key:      []byte("foo"),
+					modified: Revision{Main: 10},
+					generations: []generation{
+						{ver: 3, created: Revision{Main: 1}, revs: []Revision{Revision{Main: 5}, Revision{Main: 9}}},
+						{ver: 1, created: Revision{Main: 10}, revs: []Revision{Revision{Main: 10}}},
+					},
+				},
+				{
+					key:      []byte("foo1"),
+					modified: Revision{Main: 10, Sub: 1},
+					generations: []generation{
+						{ver: 1, created: Revision{Main: 10, Sub: 1}, revs: []Revision{Revision{Main: 10, Sub: 1}}},
+					},
+				},
+				{
+					key:      []byte("foo2"),
+					modified: Revision{Main: 8},
+					generations: []generation{
+						{ver: 3, created: Revision{Main: 3}, revs: []Revision{Revision{Main: 8}}},
+						{},
+					},
+				},
+			},
+			keep: map[Revision]struct{}{
+				Revision{Main: 5}: {},
+			},
+			compacted: map[Revision]struct{}{
+				Revision{Main: 8}: {},
+				Revision{Main: 5}: {},
+			},
+		},
+		{
+			atRev: 9,
+			keyIndexes: []keyIndex{
+				{
+					key:      []byte("foo"),
+					modified: Revision{Main: 10},
+					generations: []generation{
+						{ver: 3, created: Revision{Main: 1}, revs: []Revision{Revision{Main: 9}}},
+						{ver: 1, created: Revision{Main: 10}, revs: []Revision{Revision{Main: 10}}},
+					},
+				},
+				{
+					key:      []byte("foo1"),
+					modified: Revision{Main: 10, Sub: 1},
+					generations: []generation{
+						{ver: 1, created: Revision{Main: 10, Sub: 1}, revs: []Revision{Revision{Main: 10, Sub: 1}}},
+					},
+				},
+			},
+			keep: map[Revision]struct{}{},
+			compacted: map[Revision]struct{}{
+				Revision{Main: 9}: {},
+			},
+		},
+		{
+			atRev: 10,
+			keyIndexes: []keyIndex{
+				{
+					key:      []byte("foo"),
+					modified: Revision{Main: 10},
+					generations: []generation{
+						{ver: 1, created: Revision{Main: 10}, revs: []Revision{Revision{Main: 10}}},
+					},
+				},
+				{
+					key:      []byte("foo1"),
+					modified: Revision{Main: 10, Sub: 1},
+					generations: []generation{
+						{ver: 1, created: Revision{Main: 10, Sub: 1}, revs: []Revision{Revision{Main: 10, Sub: 1}}},
+					},
+				},
+			},
+			keep: map[Revision]struct{}{
+				Revision{Main: 10}:         {},
+				Revision{Main: 10, Sub: 1}: {},
+			},
+			compacted: map[Revision]struct{}{
+				Revision{Main: 10}:         {},
+				Revision{Main: 10, Sub: 1}: {},
+			},
+		},
 	}
+
+	ti := buildTreeIndex()
```

Comment:
Please keep the original comment
```suggestion
	ti := buildTreeIndex()
	// Continuous Compact and Keep
```

Reply:
Also put the `ti := buildTreeIndex()` inside the `for` loop?

Reply:
Updated.

> Also put the ti := buildTreeIndex() inside the for loop?

The case keeps performing compaction on one `ki`. So we can't put it inside the loop.